### PR TITLE
[Inductor max autotune] Multithreaded Precompilation

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119010
* #119009
* #119008
* #119007
* #119006
* #119004
* __->__ #119386

When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by kadeng** [comment]
This is a recreated already accepted PR https://github.com/pytorch/pytorch/pull/119003

**Review by aakhundov — APPROVED** [review]

File: torch/_inductor/autotune_process.py [hunk 1]
```diff
@@ -556,6 +565,13 @@ def __init__(
         self.source_file: str = ""
         self.hash_key, self.source_file = CUDACodeCache.write(self.source_code, "so")
 
+    def precompile(self):
+        # Prepopulate CUDACodeCache
+        # may happen in separate Threadpool
+        log.debug("Precompiling %s", str(self))
```
Comment (Skylion007) [inline torch/_inductor/autotune_process.py:571]:
```suggestion
        log.debug("Precompiling %s", self)
```
should call str o self I think?

File: torch/_inductor/autotune_process.py [hunk 2]
```diff
@@ -556,6 +565,13 @@ def __init__(
...
+        CUDACodeCache.load(self.source_code, "so")
+        log.debug("Done precompiling %s", str(self))
```
Comment (Skylion007) [inline torch/_inductor/autotune_process.py:573]:
```suggestion
        log.debug("Done precompiling %s", self)
```

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -744,7 +745,7 @@ def load(
         Load a compiled graph from the cache. If a cached entry does not exist,
         compile the graph and save it to the cache.
         """
-        from filelock import FileLock
+        from filelock import FileLock  # type: ignore[import-not-found]
```
Comment (Skylion007) [inline torch/_inductor/codecache.py:748]: We just need to add `types-filelock` to our lintrunner install.

Comment (Skylion007) [reply torch/_inductor/codecache.py:748]: Opened a PR: https://github.com/pytorch/pytorch/pull/119390

Comment (kadeng) [reply torch/_inductor/codecache.py]: thanks for the comments, I made some small corresponding changes..

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 4]
```diff
@@ -298,14 +298,18 @@ def __init__(
         layout: Layout,
         make_kernel_render: Callable[[CUDATemplateBuffer, Optional[List[IRNode]]], str],
         bmreq: CUDABenchmarkRequest,
-        template: "CUDATemplate",
+        template: "CUDATemplate",  # type: ignore[name-defined]
```
Comment (Skylion007) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:301]: Huh, this type ignore shouldn't be necessary...

Comment (kadeng) [reply torch/_inductor/codegen/cuda/cuda_kernel.py:301]: I add these type ignores only on linter errors when I run lintrunner locally. I also have the impression that these are a bit inconsistent, e.g. sometimes I had linter warnings about code that was just fine the day before.

**Pull request closed without merging**